File: backend/internal/models/websocket.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// MessageType defines the type of WebSocket message
type MessageType string

const (
	// Connection messages
	MessageTypeJoinRoom   MessageType = "join_room"
	MessageTypeLeaveRoom  MessageType = "leave_room"
	MessageTypeUserJoined MessageType = "user_joined"
	MessageTypeUserLeft   MessageType = "user_left"

	// Presence messages
	MessageTypeCursorMove      MessageType = "cursor_move"
	MessageTypeSelectionChange MessageType = "selection_change"
	MessageTypePresenceUpdate  MessageType = "presence_update"

	// Operation messages
	MessageTypeOperation MessageType = "operation"
	MessageTypeBatch     MessageType = "batch"

	// Sync messages
	MessageTypeSyncRequest  MessageType = "sync_request"
	MessageTypeSyncResponse MessageType = "sync_response"

	// Control messages
	MessageTypeHeartbeat MessageType = "heartbeat"
	MessageTypePong      MessageType = "pong"
	MessageTypeError     MessageType = "error"
)

// WSMessage represents a WebSocket message
type WSMessage struct {
	Payload   interface{} `json:"payload,omitempty"`
	Timestamp time.Time   `json:"timestamp"`
	UserID    uuid.UUID   `json:"user_id,omitempty"`
	Type      MessageType `json:"type"`
	RequestID string      `json:"request_id,omitempty"` // For request/response matching
}

// JoinRoomPayload is the payload for join_room message
type JoinRoomPayload struct {
	WorkspaceID uuid.UUID `json:"workspace_id"`
	UserColor   string    `json:"user_color,omitempty"` // Hex color for user cursor
}

// UserJoinedPayload is broadcast when a user joins
type UserJoinedPayload struct {
	UserID    uuid.UUID `json:"user_id"`
	UserName  string    `json:"user_name"`
	UserColor string    `json:"user_color"`
}

// UserLeftPayload is broadcast when a user leaves
type UserLeftPayload struct {
	UserID uuid.UUID `json:"user_id"`
}

// CursorPosition represents a user's cursor position
type CursorPosition struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// CursorMovePayload is sent when user moves cursor
type CursorMovePayload struct {
	Position CursorPosition `json:"position"`
}

// SelectionChangePayload is sent when user changes selection
type SelectionChangePayload struct {
	ElementIDs []uuid.UUID `json:"element_ids"`
}

// UserPresence represents a user's presence in the workspace
type UserPresence struct {
	UserID           uuid.UUID       `json:"user_id"`
	Cursor           *CursorPosition `json:"cursor,omitempty"`
	SelectedElements []uuid.UUID     `json:"selected_elements,omitempty"`
	LastSeen         time.Time       `json:"last_seen"`
	UserName         string          `json:"user_name"`
	UserColor        string          `json:"user_color"`
}

// PresenceUpdatePayload is broadcast to other users
type PresenceUpdatePayload struct {
	Presence UserPresence `json:"presence"`
}

// OperationType defines the type of CRDT operation
type OperationType string

const (
	OperationTypeCreate OperationType = "create"
	OperationTypeUpdate OperationType = "update"
	OperationTypeDelete OperationType = "delete"
	OperationTypeMove   OperationType = "move"
)

// OperationPayload represents a CRDT operation
type OperationPayload struct {
	ElementID   uuid.UUID     `json:"element_id"`
	WorkspaceID uuid.UUID     `json:"workspace_id"`
	UserID      uuid.UUID     `json:"user_id"`
	Data        interface{}   `json:"data,omitempty"` // Element data for create/update
	Timestamp   int64         `json:"timestamp"`      // Lamport timestamp
	OpType      OperationType `json:"op_type"`
}

// BatchPayload contains multiple operations
type BatchPayload struct {
	Operations []OperationPayload `json:"operations"`
}

// SyncRequestPayload requests synchronization
type SyncRequestPayload struct {
	WorkspaceID uuid.UUID        `json:"workspace_id"`
	StateVector map[string]int64 `json:"state_vector"` // user_id -> last_seen_timestamp
}

// SyncResponsePayload contains operations to sync
type SyncResponsePayload struct {
	StateVector map[string]int64   `json:"state_vector"` // Current state vector
	Operations  []OperationPayload `json:"operations"`
}

// ErrorPayload represents an error message
type ErrorPayload struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// Client represents a connected WebSocket client
type Client struct {
	ID          uuid.UUID
	UserID      uuid.UUID
	WorkspaceID uuid.UUID
	Presence    *UserPresence
	Send        chan *WSMessage // Channel for outbound messages
	LastPing    time.Time
	UserName    string
	UserColor   string
}

// Room represents a workspace collaboration room
type Room struct {
	WorkspaceID uuid.UUID
	Clients     map[uuid.UUID]*Client // client_id -> client
	Broadcast   chan *WSMessage       // Broadcast channel
	Register    chan *Client          // Register channel
	Unregister  chan *Client          // Unregister channel
}

File: frontend/src/lib/types/websocket.ts
// WebSocket Message Types for Real-time Collaboration

export type MessageType =
	// Connection
	| 'join_room'
	| 'leave_room'
	| 'user_joined'
	| 'user_left'
	// Presence
	| 'cursor_move'
	| 'selection_change'
	| 'presence_update'
	// Operations (CRDT)
	| 'operation'
	| 'batch'
	// Sync
	| 'sync_request'
	| 'sync_response'
	// Control
	| 'heartbeat'
	| 'pong'
	| 'error';

export type OperationType = 'create' | 'update' | 'delete' | 'move';

export interface WSMessage<T = unknown> {
	type: MessageType;
	payload?: T;
	timestamp: string; // ISO string
	user_id?: string;
	request_id?: string; // For request/response pattern
}

// Connection Payloads

export interface JoinRoomPayload {
	workspace_id: string;
	user_color?: string; // Hex color for cursor
}

export interface LeaveRoomPayload {
	workspace_id: string;
}

export interface UserJoinedPayload {
	user_id: string;
	user_name: string;
	user_color: string;
	cursor?: CursorPosition;
}

export interface UserLeftPayload {
	user_id: string;
}

// Presence Payloads

export interface CursorPosition {
	x: number;
	y: number;
}

export interface CursorMovePayload {
	workspace_id: string;
	cursor: CursorPosition;
}

export interface SelectionChangePayload {
	workspace_id: string;
	selected_elements: string[]; // Element IDs
}

export interface UserPresence {
	user_id: string;
	user_name: string;
	user_color: string;
	cursor?: CursorPosition;
	selected_elements: string[];
	last_seen: string; // ISO timestamp
}

export interface PresenceUpdatePayload {
	presence: UserPresence;
}

// Operation Payloads (CRDT)

export interface OperationPayload {
	element_id: string;
	workspace_id: string;
	user_id: string;
	op_type: OperationType;
	data?: Record<string, unknown>; // Element data or partial update
	timestamp: number; // Lamport clock
}

export interface BatchPayload {
	operations: OperationPayload[];
}

// Sync Payloads

export interface SyncRequestPayload {
	workspace_id: string;
	state_vector: Record<string, number>; // user_id -> last_timestamp
}

export interface SyncResponsePayload {
	workspace_id: string;
	state_vector: Record<string, number>;
	operations: OperationPayload[];
}

// Control Payloads

export interface ErrorPayload {
	code: string;
	message: string;
	details?: string;
}

// WebSocket Connection State

export type ConnectionState =
	| 'connecting'
	| 'connected'
	| 'disconnected'
	| 'reconnecting'
	| 'error';

export interface ConnectionStatus {
	state: ConnectionState;
	error?: string;
	reconnectAttempt?: number;
	maxReconnectAttempts?: number;
}
